### PR TITLE
Auto completion from text buffer

### PR DESCRIPTION
**Pull request by lochrist**
- Auto complete stingray specific API from documentation extracted from Adoc.
- Implement IdString evaluation in debugger using the strings.txt file. Uses same parsing techniques as atom-stingray. Disabled for now until we have a proper way of initializing the projectPath.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
 	"typescript.tsdk": "node_modules/typescript/lib",
 	"files.trimTrailingWhitespace": true,
 	"eslint.enable": false,
-	"editor.insertSpaces": false,
+	"editor.insertSpaces": true,
```
Comment (lochrist) [inline .vscode/settings.json]: Should we have those settings in the extension?

File: src/idstring.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import * as fs from 'fs';
```
Comment (lochrist) [inline src/idstring.ts]: Taken from atom-stingray

File: src/murmurhash.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
+function uint64(hi, lo) {
```
Comment (lochrist) [inline src/murmurhash.ts]: Taken from atom-stingray

File: src/stingray-debugger.ts [hunk 4]
```diff
@@ -165,6 +170,12 @@ class StingrayDebugSession extends DebugSession {
         // This debugger uses one-based lines and columns
         this.setDebuggerLinesStartAt1(true);
         this.setDebuggerColumnsStartAt1(true);
+
+        /*
+        // Where is the good place to init this? And where to get the project path?
```
Comment (lochrist) [inline src/stingray-debugger.ts]: I wasn't sure where we could initialize idString. When we get a callstack? This means we can only get it when a breakpoint is triggered... could we know the project folder before? as soon as we connect to the engine?

File: src/stingray-debugger.ts [hunk 5]
```diff
@@ -725,6 +747,42 @@ class StingrayDebugSession extends DebugSession {
         return request.promise;
     }
 
+    private initIdString(projectPath : string): void {
+        if (!fs.existsSync(projectPath) || fs.exists(path.join(projectPath, '.stingray_project'))) {
+            throw new Error("Project path doesn't exists or is invalid");
```
Comment (lochrist) [inline src/stingray-debugger.ts]: should we throw when we have error like these? What would be our accepted method of reporting errors?

File: src/autocomplete.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+class Adoc {
+    private content: any = null;
+
+    constructor(adocFile) {
+        if (!fs.existsSync(adocFile)) {
+            throw new Error("Adoc file doesn't exist: ");
+        }
+
+        let buffer = fs.readFileSync(adocFile);
```
Comment (lochrist) [inline src/autocomplete.ts]: All readFileSync I have used are problematic when debugging. it seems if you F10 over a big file read like this, that the debugger is hanging. But the reading still happens. Just don't break around readFileSync.

File: src/stingray-debugger.ts [hunk 7]
```diff
@@ -725,6 +747,42 @@ class StingrayDebugSession extends DebugSession {
...
+            throw new Error('Cannot find project data path:' + projectDataPath);
+        }
+
+        let idStringPath = path.join(projectDataPath, 'win32', 'strings.txt');
+        if (!fs.existsSync(idStringPath)) {
+            throw new Error('Cannot find id string file:' + idStringPath);
+        }
+
+        try {
+            let buffer = fs.readFileSync(idStringPath);
```
Comment (lochrist) [inline src/stingray-debugger.ts]: Same as above. Tends to make the debugger hang.

File: src/stingray-debugger.ts [hunk 8]
```diff
@@ -725,6 +747,42 @@ class StingrayDebugSession extends DebugSession {
...
+        }
+    }
+
+    private isIdStringExpression(expression: string) : boolean {
+        return expression.length > 0 && expression.indexOf('#ID[') === 0 && expression[expression.length - 1] === ']';
+    }
+
+    private evaluateIdString(expression: string) : string {
+        let idString = expression.substr(4, expression.length - 5);
+        let str = this._idStringMap[idString];
```
Comment (lochrist) [inline src/stingray-debugger.ts]: this._idStringMap is not accesible from the debugger.

File: .vscode/settings.json (hunk 1 above)
Comment (jschmidt42) [inline .vscode/settings.json]: Yes!

**Comment by jschmidt42** [comment]
Looks good. Merge that in and I'll create a new version.

**Pull request merged**